src/__tests__/old-act.js

```diff
@@ -4,32 +4,92 @@ jest.mock('../react-dom-16.9.0-is-released', () => ({
   reactDomSixteenPointNineIsReleased: true,
 }))
 
-jest.mock('react-dom/test-utils', () => ({
-  act: cb => {
-    const promise = cb()
-    return {
-      then() {
-        console.error('blah, do not do this')
-        return promise
-      },
-    }
-  },
-}))
-
 test('async act works even when the act is an old one', async () => {
+  jest.mock('react-dom/test-utils', () => ({
+    act: cb => {
+      cb()
+      return {
+        then() {
+          console.error(
+            'Warning: Do not await the result of calling ReactTestUtils.act(...), it is not a Promise.',
+          )
+        },
+      }
+    },
+  }))
+
   jest.spyOn(console, 'error').mockImplementation(() => {})
   const callback = jest.fn()
   await asyncAct(async () => {
     await Promise.resolve()
     await callback()
   })
   expect(console.error.mock.calls).toMatchInlineSnapshot(`
-Array [
-  Array [
-    "It looks like you're using a version of react-dom that supports the \\"act\\" function, but not an awaitable version of \\"act\\" which you will need. Please upgrade to at least react-dom@16.9.0 to remove this warning.",
-  ],
-]
-`)
+    Array [
+      Array [
+        "It looks like you're using a version of react-dom that supports the \\"act\\" function, but not an awaitable version of \\"act\\" which you will need. Please upgrade to at least react-dom@16.9.0 to remove this warning.",
+      ],
+    ]
+  `)
+  expect(callback).toHaveBeenCalledTimes(1)
+
+  // and it doesn't warn you twice
+  callback.mockClear()
+  console.error.mockClear()
+
+  await asyncAct(async () => {
+    await Promise.resolve()
+    await callback()
+  })
+  expect(console.error).toHaveBeenCalledTimes(0)
+  expect(callback).toHaveBeenCalledTimes(1)
+
+  console.error.mockRestore()
+
+  jest.unmock('react-dom/test-utils')
+})
+
+test('should not warn with a version that resolves', async () => {
+  jest.mock('react-dom/test-utils', () => ({
```

Comment:
I'm pretty sure that putting `jest.mock` and `jest.unmock` throughout this file would require that you do `jest.resetModules()` and `require` the modules in each test... 🤔

I do it that way sometimes, but the way I normally solve this problem is by putting the tests that require different mocks in separate files or by making the thing I want mocked a mock function and using `mockImplementationOnce` for it.

In this case, one file for old react (that doesn't have act) and another for one that does which behaves differently (via `mockImplementationOnce`) seems like the right solution.

Reply:
yeah I might not be doing the right thing with these tests. mockImplementationOnce is a good lead.